step-15 Accessibility

felixzapata opened the pull request:
- Add ngAria module.
- Add labels to the search and order fields.
- Add accessibility plugin for Protractor.
  - Add missing alt attributes in the phone detail.
- Add aria live regions to inform the user about results after searching and filtering elements.
-  Improve access via keyboard:
  - Navigate between the images in a phone detail.
  - Add headings elements.
- Add information about checkmarks in the phone detail specs.

By the way, do I have to modify the document in the [full tutorial](https://github.com/angular/angular.js/tree/master/docs/content/tutorial)?

felixzapata commented:
hi @gkalpak, ¿is there any way that the use of the `debounce` option for the `ng-model` works with Protractor?  I need this option for my changes. But with that, the e2e tests fail.

I know that there is an [issue about that](https://github.com/angular/protractor/issues/1671) but nothing more.

gkalpak commented:
Without having really looked into that, I can't imagine how protractor can be ignoring the debounce. It's something that Angular does and PRotractor doesn't have a way to "ignore" it.

What is probably happening, is that Protractor waits for the view to "stabilize" (which in Angular terms includes waiting for registered timeouts, such as the one used by `debounce`) and only then assets your expectation.

If you want to circumvent that, you can turn on `ignoreSynchronization` (read more about it [here][1]).

--
If you can't get it to work, post the failing test and I'll take a look.

[1]: https://github.com/angular/protractor/blob/master/docs/timeouts.md

felixzapata commented:
I think that with `ignoreSynchronization` does not work. I saw yesterday this information about timeouts. One of the test that fails is:

```javascript
    it('should filter the phone list as a user types into the search box', function() {
      var phoneList = element.all(by.repeater('phone in $ctrl.phones'));
      var query = element(by.model('$ctrl.query'));

      expect(phoneList.count()).toBe(20);

      query.sendKeys('nexus');
      expect(phoneList.count()).toBe(1);

      query.clear();
      query.sendKeys('motorola');
      expect(phoneList.count()).toBe(8);
    });
```
The problem is the `expect` runs before the `debounce` finishes, so the results never will be the correct. It Is the same example that an user reported in the Protractor issue that I mentioned before.

I have tried using `wait` and `setTimeout` but I obtain the same errors.

felixzapata commented:
@gkalpak After a review, I think that my problem is related with some changes I've made in the repeater. I will check it.

felixzapata commented:
@gkalpak done. Now the e2e tests pass in my local environment. I've added an `browser.sleep` suggested by @mohitjee15 and I've changed how to obtain the repeater in the scenario due to my changes in the template.

gkalpak commented:
Thx :+1:
I will take a look this week !

felixzapata commented:
ok @gkalpak. Do I have to modify the  [full tutorial](https://github.com/angular/angular.js/tree/master/docs/content/tutorial) and send another pull request or I have to wait until this pull request is accepted?

gkalpak commented:
Yes, you have to update the tutorial context as well, so we can merge them together.

felixzapata commented:
ok, I will work in the tutorial and I will send another pulll request as soon as possible.

gkalpak commented:
@felixzapata, I'll review this together with the tutorial content PR - feel free to ping me when it's ready.
Thx for working on this :+1:

felixzapata commented:
@gkalpak I've created the [pull request with the new step in the tutorial](https://github.com/angular/angular.js/pull/14779). 

About this pull request, please feel free to comment if I need to review the code in the new directives due to some mistakes.

gkalpak commented:
Awesome! I will take a look soon.

gkalpak commented:
Note to self: We need a new tag for this step.

gkalpak commented:
Why had the previous steps be modified? Since this is an extra step at the end, previous steps should not be touched as far as I understand. Do I miss something?

felixzapata commented:
@gkalpak I will check again the pull request. I'm only want to upload three commits and I though that it was ok.

felixzapata commented:
@gkalpak  done, a problem with the rebase

gkalpak commented on package.json:
```diff
@@ -13,7 +13,9 @@
-    "protractor": "^3.2.2"
+    "protractor": "^3.2.2",
+    "protractor-accessibility-plugin": "^0.1.1",
+    "shelljs": "^0.6.0"
```
Is this necessary? It doesn't seem to be used.

gkalpak commented on README.md:
```diff
@@ -165,6 +165,19 @@ git diff step-?..step-?
+  
+### step-15 _Accessibility (a11y)_
+
+- Add ngAria module.
```
Didn't you end up removing it in the end? (Although it does seem strange to have a chapter about ARIA that doesn't mention or need `ngAria` :smiley:)

felixzapata replied:
Yes, finally I removed the module so I will update the README file.

felixzapata commented:
@gkalpak I´ve updated the PR to avoid the multiple commits. 

I will take a look to your comments.

felixzapata commented:
@gkalpak is there a problem inside Travis with the driver for Chrome?

gkalpak commented:
Not sure, but if the tests pass locally, don't worry about Travis. I'll deal with it later.
